src/Pelagos/Util/DOIutil.php

```diff
@@ -375,4 +334,51 @@ function ($matches) {
             $input
         );
     }
+
+    /**
+     * This function generates a new DOI string that is not already in use per datacite.
+     *
+     * @return string A unique, unused DOI string.
+     */
+    public function generateDoi()
+    {
+        $encoder = new Base32CrockfordEncoder();
+        // 1099511627775 encodes to the longest 8 character Crockford 32 string.
+        $max = 1099511627775;
+        // Start at 1 (0 is problematic as library does not produce checksum for 0.)
+        $random = mt_rand(1, $max);
+        // Add shoulder and remove the checksum character.
+        $doi = $this->doishoulder . '/' . substr($encoder->encode($random), 0, -1);
+        while ($this->checkDoiExistsExternal($doi)) {
+            $random = mt_rand(1, $max);
+            $doi = $this->doishoulder . '/' . substr($encoder->encode($random), 0, -1);
+        }
+        return $doi;
+    }
+
+    /**
+     * This function queries Datacite to determine if the DOI is already in use.
+     *
+     * @param string $doi The DOI to be checked.
+     *
+     * @throws HttpClientErrorException If a non 400 error is encountered.
+     *
+     * @return boolean True if returned if the DOI exists, false if it does not.
+     */
+    public function checkDoiExistsExternal(string $doi): bool
+    {
+        try {
+            $doiUtil = new DOIutil();
+            $doiUtil->getDOIMetadata($doi);
+        } catch (HttpClientErrorException $exception) {
+            if ($exception->getCode() == 400) {
```

Comment:
Will it be enough to just create a new DOI, if we get a Client side exception. I've not seen any other 4xx codes returned by the GET API

Reply:
I think the more proper response should have been a 404 from datacite, and is case they do this is the future, I'll check for both 400 and 404. If they ever do correct this in the future, we should probably re-throw the 400.

Reply:
updated to look for either 404 or 400 here.